nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py

```diff
@@ -0,0 +1,290 @@
+            self.bias = Parameter(torch.empty(self.module.bias.size()))
+            self.bias.data = self.module.bias.data
+        else:
+            self.register_buffer("bias_mask", None)
+
+    def _weight2buffer(self):
+        """
+        When using this wrapper to inference, call `_weight2buffer()` to make original weight untrainable.
+        The best place to call this function is in `Pruner._wrap_model()`.
+        """
+        delattr(self.module, 'weight')
+        self.module.register_buffer('weight', self.weight.data)
+        if hasattr(self.module, 'bias') and self.module.bias is not None:
+            delattr(self.module, 'bias')
+            self.module.register_buffer('bias', self.bias.data)
+
+    def _weight2parameter(self):
+        """
+        When don't need to record score or need to export the model, call `_weight2parameter()` to make the original weight trainable.
+        The best place to call this function is in `Pruner._unwrap_model()`.
+        """
+        delattr(self.module, 'weight')
+        self.module.weight = Parameter(torch.empty(self.weight.size()))
+        self.module.weight.data = torch.mul(self.weight, self.weight_mask)
+        if hasattr(self.module, 'bias') and self.module.bias is not None:
+            delattr(self.module, 'bias')
+            self.module.bias = Parameter(torch.empty(self.bias.size()))
+            self.module.bias.data = torch.mul(self.bias, self.bias_mask)
+
+    def forward(self, *inputs):
+        # apply mask to weight, bias
+        self.module.weight = torch.mul(self.weight, _StraightThrough.apply(self.weight_score, self.weight_mask))
+        if hasattr(self.module, 'bias') and self.module.bias is not None:
+            self.module.bias = torch.mul(self.bias, self.bias_mask)
+        return self.module(*inputs)
+
+
+class _StraightThrough(autograd.Function):
+    """
+    Straight through the gradient to the score, then the score = initial_score + sum(-lr * grad(weight) * weight).
+    """
+    @staticmethod
+    def forward(self, score, masks):
+        return masks
+
+    @staticmethod
+    def backward(ctx, gradOutput):
+        return gradOutput, None
+
+
+class WeightScoreTrainerBasedDataCollector(TrainerBasedDataCollector):
+    """
+    Collect all weight_score in wrappers as data used to calculate metrics.
+    """
+    def _reset_optimizer(self):
+        """
+        Weed out the weight_score from the parameters passed to optimizer, guaranteed to load the optimizer state dict.
+        """
+        if self._origin_optimizer_cls is not None:
+            optimizer_grouped_parameters = [{
+                "params": [p for n, p in self.compressor.bound_model.named_parameters() if "weight_score" not in n and p.requires_grad]
+            }]
+            if self._origin_optimizer_cls.__name__ == 'SGD':
+                self.optimizer = self._origin_optimizer_cls(optimizer_grouped_parameters, lr=0.001)
```

Comment:
why hardcode 0.001 here?

Reply:
Because lr is a required parameter in `SGD`, and `0.001` is a fake number, it will be reload by `self.optimizer.load_state_dict(self._origin_optimizer_state_dict)`